executor/join.go

```diff
@@ -296,32 +319,105 @@ func (e *HashJoinExec) fetchInnerRows(ctx context.Context, chkCh chan<- *chunk.C
 	}
 }
 
-// evalRadixBitNum evaluates the radix bit numbers.
-func (e *HashJoinExec) evalRadixBitNum() {
-	sv := e.ctx.GetSessionVars()
-	// Calculate the bit number needed when using radix partition.
-	if !sv.EnableRadixJoin {
-		return
+// partitionInnerRows re-order e.innerResults into sub-relations.
+func (e *HashJoinExec) partitionInnerRows() error {
+	if err := e.preAlloc4InnerParts(); err != nil {
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	defer wg.Wait()
+	wg.Add(int(e.concurrency))
+	for i := 0; i < int(e.concurrency); i++ {
+		workerID := i
+		go util.WithRecovery(func() {
+			e.doInnerPartition(workerID, wg)
+			wg.Done()
+		}, nil)
+	}
+	return nil
+}
+
+// doInnerPartition runs concurrently, partitions and copies the inner relation
+// to several pre-allocated data partitions. The input inner Chunk idx for each
+// partitioner is workerId + x*numPartitioners.
+func (e *HashJoinExec) doInnerPartition(workerID int, wg *sync.WaitGroup) {
+	chkIdx, chkNum := workerID, e.innerResult.NumChunks()
+	for ; chkIdx < chkNum; chkIdx += int(e.concurrency) {
+		chk := e.innerResult.GetChunk(chkIdx)
+		for srcRowIdx, partPtr := range e.innerRowPrts[chkIdx] {
+			partIdx, destRowIdx := partPtr.partitionIdx, partPtr.rowIdx
+			part := e.innerParts[partIdx]
+			part.Insert(int(destRowIdx), chk.GetRow(srcRowIdx))
+		}
 	}
+}
+
+// preAlloc4InnerParts evaluates partRowPtr and pre-alloc the memory space
+// for every inner row to help re-order the inner relation.
+// TODO: we need to evaluate the skewness for the partitions size, if the
+// skewness exceeds a threshold, we do not use partition phase.
+func (e *HashJoinExec) preAlloc4InnerParts() (err error) {
+	hasNull, keyBuf := false, make([]byte, 0, 64)
+	for chkIdx, chkNum := 0, e.innerResult.NumChunks(); chkIdx < chkNum; chkIdx++ {
+		chk := e.innerResult.GetChunk(chkIdx)
+		partPtrs := make([]partRowPtr, chk.NumRows())
+		for rowIdx := 0; rowIdx < chk.NumRows(); rowIdx++ {
+			row := chk.GetRow(rowIdx)
+			hasNull, keyBuf, err = e.getJoinKeyFromChkRow(false, row, keyBuf)
+			if err != nil {
+				return err
+			}
+			if hasNull {
+				continue
```

Comment:
@zz-jason Seems not.
If the elements in a row are all NULL, it will not be inserted into a partition.
If part of the elements in a row are NULL, no errors would happen.